types/context.go

```diff
@@ -302,6 +303,18 @@ func (c Context) CacheContext() (cc Context, writeCache func()) {
 	return cc, writeCache
 }
 
+// RunAtomic runs the fn atomically, if fn returns error or panic, partial state won't be saved.
+func (c Context) RunAtomic(fn func(Context) error) error {
+	cacheCtx := c.WithCacheMultiStore(c.cms.Clone()).WithEventManager(NewEventManager())
+	if err := fn(cacheCtx); err != nil {
+		return err
+	}
+
+	c.cms.Restore(cacheCtx.cms)
+	c.EventManager().EmitEvents(cacheCtx.EventManager().Events())
+	return nil
```

Comment:
This isn't safe: it's possible for another caller to interact with the context's cms between the fn call and the Restore call in this method, and in that case the mutations of that caller would be overwritten and lost. Equivalently, it's possible for another caller to write to the context's cms after this method calls Restore but before it returns, which would overwrite and lose the mutations of the RunAtomic caller. So this basically doesn't work.

Reply:
Consensus state machine runs sequentially, that's the implicit assumption here.

Reply:
A context is not guaranteed to be used exclusively in a single-threaded manner by the consensus state machine, right? A context can absolutely be accessed by concurrent callers.

Reply:
> A context is not guaranteed to be used exclusively in a single-threaded manner by the consensus state machine, right? A context can absolutely be accessed by concurrent callers.

I think that's the case in current sdk and our chain, but I can't speak for the other chains though.

Reply:
It doesn't matter how a package happens to be used by consumers right now, what matters is how the package can possibly be used by any consumer in the future. If a package has a Foo method that takes an int, and all current consumers happen to call Foo with a value of 0, that doesn't mean that Foo can assume it will be called with a value of 0, it has to deal with any valid int. Similarly just because a context happens to be called in a single threaded way by the chains you care about, that doesn't mean that the context can assume that it will always be called in a single threaded way by all consumers.

Reply:
No caller can decide if his use case concurrent or not, while the callback work on a branched out ctx, it's ok for the callback to do concurrency inside. So for example, it's safe to call this function in runTx, while it's ok for the message handler to do concurrency inside, does that make sense?

Reply:
> the callback work on a branched out ctx

That branched context is the result of calling c.cms.Clone, but if the "atomic" transaction succeeds, you then call c.cms.Restore, which mutates the original c.cms value. And because c.cms is an interface, and therefore has reference semantics, the Restore call can happen concurrently with any other operation on that cms value. But CacheMultiStore doesn't say that its implementations need to be safe for concurrent use. So any other goroutine with a context that has the same CacheMultiStore can potentially do a read while that Restore is doing a write. That's a data race, and as long as the CacheMultiStore is not documented as being safe for concurrent use, preventing that data race is the responsibility of the context and its methods.

> it's ok for the callback to do concurrency inside

As the type exists today, no, it actually isn't.

The CacheMultiStore type doesn't document that it is safe for concurrent use. That means that any code which uses a CacheMultiStore, and can be called concurrently, must serialize access to those stores.

An SDK context's methods can be called concurrently. This isn't a point of debate: the context type is exported, and its methods are part of the SDK's public API. That means callers can call its types and methods however they like. If the context type were documented to make it clear that it isn't safe for concurrent use, that would be a different story, of course. But that documentation doesn't exist, and so that assumption can't be assumed.

Reply:
AFAIK, sdk has rarely made it clear in doc if a method if thread-safe or not, our assumption has always been defaulting to unsafe when unclear.
`RunAtomic` is not thread-safe when calling on the same ctx, no doubt about that, so the issue is improving the doc to make that clearer, or do you think we shouldn't put it as a context method?